pkg/subsystem/service.go

```diff
@@ -0,0 +1,44 @@
+// Copyright 2023 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package subsystem
+
+import (
+	"fmt"
+)
+
+type Service struct {
+	*Extractor
+	perName map[string]*Subsystem
+}
+
+func MustMakeService(list []*Subsystem) *Service {
+	service, err := MakeService(list)
+	if err != nil {
+		panic(fmt.Sprintf("service creation failed: %s", err))
+	}
+	return service
+}
+
+func MakeService(list []*Subsystem) (*Service, error) {
+	extractor := MakeExtractor(list)
+	perName := map[string]*Subsystem{}
+	for _, item := range list {
+		if item.Name == "" {
+			return nil, fmt.Errorf("input contains a subsystem without a name")
+		}
+		if perName[item.Name] != nil {
+			return nil, fmt.Errorf("collision on %#v name", item.Name)
+		}
+		perName[item.Name] = item
```

Comment:
All of this code makes me think: if indexing by name is what we want from subsystems in the end, why don't we store them in the map in the first place? Map also allows iteration, so it's not worse in other regards.
groupByList could accept []maintainersRecord, do all of the grouping and naming (including custom names as we discussed) and from then on we only have subsystems in map[string]*Subsystem.

Reply:
During the subsystem generation, we'd like to delay name assignment to a later moment. The fewer subsystems are remaining, the better are our chances at generating a good name for them (fewer duplicates and in general fewer names to generate).

While saving to `lists/*.go`, I think we could do it as a map, though that's the question of saving only a few lines of code in `service.go`.